### PR TITLE
[expo-updates][ios] Add support for multipart manifest responses

### PR DESCRIPTION
**Pull request by wschurman**
# Why

iOS equivalent of https://github.com/expo/expo/pull/15401/. The same How and Test Plan apply, but on iOS.

# How

- `EXUpdatesParameterParser` is a translation to objective-c of ParameterParser from https://commons.apache.org/proper/commons-fileupload/. The test for it is a combination of their test cases and ours.
- `EXUpdatesMultipartStreamReader` is a fork of `RCTMultipartStreamReader` that fixes a bug in preamble skipping and removes features we don't need. A new test was added for it to test the preamble case, as well as all the other cases that were already tested.

# Test Plan

Run tests.

### CODE REVIEW

File: packages/expo-updates/ios/EXUpdates/AppLoader/EXUpdatesFileDownloader.m [hunk 1]
```diff
@@ -85,115 +88,280 @@ - (NSURLRequest *)createManifestRequestWithURL:(NSURL *)url extraHeaders:(nullab
...
+- (void)parseManifestBodyData:(NSData *)data
+             headerDictionary:(NSDictionary *)headerDictionary
+                   extensions:(NSDictionary *)extensions
+                     database:(EXUpdatesDatabase *)database
+                 successBlock:(EXUpdatesFileDownloaderManifestSuccessBlock)successBlock
+                   errorBlock:(EXUpdatesFileDownloaderErrorBlock)errorBlock {
+  id headerSignature = headerDictionary[@"expo-manifest-signature"];
+  
+  NSError *err;
+  id parsedJson = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:&err];
```
Comment (jkhales) [inline packages/expo-updates/ios/EXUpdates/AppLoader/EXUpdatesFileDownloader.m:251]:
```suggestion
  id manifestBodyJson = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:&err];
```

File: packages/expo-updates/ios/EXUpdates/AppLoader/EXUpdatesFileDownloader.m [hunk 2]
```diff
@@ -85,115 +88,280 @@ - (NSURLRequest *)createManifestRequestWithURL:(NSURL *)url extraHeaders:(nullab
...
+
+  return [self parseManifestBodyData:manifestData
+                    headerDictionary:manifestHeaders
+                          extensions:extensions
+                            database:database
+                        successBlock:successBlock
+                          errorBlock:errorBlock];
+}
+
+- (void)parseManifestBodyData:(NSData *)data
```
Comment (jkhales) [inline packages/expo-updates/ios/EXUpdates/AppLoader/EXUpdatesFileDownloader.m:242]:
```suggestion
- (void)parseManifestBodyData:(NSData *)manifestBodyData
```

**Review by jkhales — APPROVED** [review]
🧑‍🚒

File: packages/expo-updates/ios/EXUpdates/AppLoader/EXUpdatesRemoteAppLoader.m [hunk 3]
```diff
@@ -97,11 +97,15 @@ - (void)downloadAsset:(EXUpdatesAsset *)asset
         return;
       }
 
-      [self->_downloader downloadFileFromURL:asset.url toPath:[urlOnDisk path] successBlock:^(NSData *data, NSURLResponse *response) {
+      [self->_downloader downloadFileFromURL:asset.url
+                                      toPath:[urlOnDisk path]
+                                extraHeaders:asset.extraRequestHeaders ?: [NSDictionary new]
```
Comment (esamelson) [inline packages/expo-updates/ios/EXUpdates/AppLoader/EXUpdatesRemoteAppLoader.m:102]:
```suggestion
                                extraHeaders:asset.extraRequestHeaders ?: @{}
```

File: packages/expo-updates/ios/EXUpdates/Database/EXUpdatesDatabase.m [hunk 4]
```diff
@@ -164,11 +165,12 @@ - (void)updateAsset:(EXUpdatesAsset *)asset error:(NSError ** _Nullable)error
   NSAssert(asset.filename, @"asset filename should be nonnull");
   NSAssert(asset.contentHash, @"asset contentHash should be nonnull");
 
-  NSString * const assetUpdateSql = @"UPDATE \"assets\" SET \"headers\" = ?2, \"type\" = ?3, \"metadata\" = ?4, \"download_time\" = ?5, \"relative_path\" = ?6, \"hash\" = ?7, \"url\" = ?8 WHERE \"key\" = ?1;";
+  NSString * const assetUpdateSql = @"UPDATE \"assets\" SET \"headers\" = ?2, \"extra_request_headers\" = ?2, \"type\" = ?3, \"metadata\" = ?4, \"download_time\" = ?5, \"relative_path\" = ?6, \"hash\" = ?7, \"url\" = ?8 WHERE \"key\" = ?1;";
```
Comment (esamelson) [inline packages/expo-updates/ios/EXUpdates/Database/EXUpdatesDatabase.m:168]:
```suggestion
  NSString * const assetUpdateSql = @"UPDATE \"assets\" SET \"headers\" = ?2, \"extra_request_headers\" = ?3, \"type\" = ?4, \"metadata\" = ?5, \"download_time\" = ?6, \"relative_path\" = ?7, \"hash\" = ?8, \"url\" = ?9 WHERE \"key\" = ?1;";
```

File: packages/expo-updates/ios/EXUpdates/Update/EXUpdatesNewUpdate.m [hunk 5]
```diff
@@ -12,13 +12,16 @@
 @implementation EXUpdatesNewUpdate
 
 + (EXUpdatesUpdate *)updateWithNewManifest:(EXManifestsNewManifest *)manifest
-                                  response:(nullable NSURLResponse *)response
+                                   headers:(NSDictionary *)headersDictionary
```
Comment (esamelson) [inline packages/expo-updates/ios/EXUpdates/Update/EXUpdatesNewUpdate.m:15]:
```suggestion
                                   headers:(NSDictionary *)headers
```
probably not a big deal, but I would keep the parameter names the same in the header and implementation files

File: packages/expo-updates/ios/EXUpdates/Update/EXUpdatesUpdate.m [hunk 6]
```diff
@@ -63,22 +63,12 @@ + (instancetype)updateWithId:(NSUUID *)updateId
 }
 
 + (instancetype)updateWithManifest:(NSDictionary *)manifest
-                          response:(nullable NSURLResponse *)response
+                           headers:(NSDictionary *)headerDictionary
```
Comment (esamelson) [inline packages/expo-updates/ios/EXUpdates/Update/EXUpdatesUpdate.m:66]:
```suggestion
                           headers:(NSDictionary *)headers
```
same comment here

File: packages/expo-updates/ios/Tests/EXUpdatesDatabaseInitializationTests.m [hunk 7]
```diff
@@ -590,4 +590,46 @@ - (void)testMigration6To7
   XCTAssertEqual(0, [EXUpdatesDatabaseUtils executeSql:selectDeletedSql3 withArgs:nil onDatabase:migratedDb error:nil].count);
 }
 
+- (void)testMigration7To8 {
+  sqlite3 *db;
+  NSError *initializeError;
+  [EXUpdatesDatabaseInitialization initializeDatabaseWithSchema:EXUpdatesDatabaseV6Schema
```
Comment (esamelson) [inline packages/expo-updates/ios/Tests/EXUpdatesDatabaseInitializationTests.m:596]:
```suggestion
  [EXUpdatesDatabaseInitialization initializeDatabaseWithSchema:EXUpdatesDatabaseV7Schema
```
the way we've been doing this for now has been just copying the old schemas into constants at the top of this file. definitely open to better ideas if you want to change this 🤷‍♂️ but either way would be good to start this test with the correct schema.

File: packages/expo-updates/ios/EXUpdates/AppLauncher/EXUpdatesAppLauncherWithDatabase.m [hunk 8]
```diff
@@ -298,7 +301,8 @@ - (void)_downloadAsset:(EXUpdatesAsset *)asset
         asset.downloadTime = [NSDate date];
         completion(nil, asset, assetLocalUrl);
       });
-    } errorBlock:^(NSError *error, NSURLResponse *response) {
+    }
+                              errorBlock:^(NSError *error) {
```
[the remark this answers is not in the document]
Comment (wschurman) [reply packages/expo-updates/ios/EXUpdates/AppLauncher/EXUpdatesAppLauncherWithDatabase.m:305]: when on a separate line xcode auto-format has this way out in space. since all the other selector args are on their own line, I prefer to have this on it's own line. And I prefer to use xcode auto-format as a source of truth (which would be awesome if we could have a lint rule for).

File: packages/expo-updates/ios/EXUpdates/Multipart/EXUpdatesParameterParser.m [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+  int _pos;
+  
+  // Maximum position in the string.
+  NSUInteger _len;
+  
+  // Start of a token.
+  int _i1;
+  
+  // End of a token.
+  int _i2;
```
[the remark this answers is not in the document]
Comment (wschurman) [reply packages/expo-updates/ios/EXUpdates/Multipart/EXUpdatesParameterParser.m:22]: yep, these are a close fork so keeping variable names makes it easier to check

File: packages/expo-updates/ios/EXUpdates/Multipart/EXUpdatesMultipartStreamReader.m [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    NSString *key = [line substringToIndex:location];
+    NSString *value = [[line substringFromIndex:location + 1] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
+    [headers setValue:value forKey:key];
+  }
+  return headers;
+}
+
+- (void)emitChunk:(NSData *)data headers:(NSDictionary *)headers callback:(RCTMultipartCallback)callback done:(BOOL)done
+{
+  NSData *marker = [CRLF CRLF dataUsingEncoding:NSUTF8StringEncoding];
```
Comment (ide) [inline packages/expo-updates/ios/EXUpdates/Multipart/EXUpdatesMultipartStreamReader.m:40]: i see what this does but lol

File: packages/expo-updates/ios/EXUpdates/Multipart/EXUpdatesMultipartStreamReader.m [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+  
+  NSData *closeDelimiter = [[NSString stringWithFormat:@"%@--%@--%@", CRLF, _boundary, CRLF] dataUsingEncoding:NSUTF8StringEncoding];
+  NSMutableData *content = [[NSMutableData alloc] initWithCapacity:1];
+  NSDictionary *currentHeaders = nil;
+  NSUInteger currentHeadersLength = 0;
+
+  const NSUInteger bufferLen = 4 * 1024;
+  uint8_t buffer[bufferLen];
+
+  [_stream open];
```
Comment (ide) [inline packages/expo-updates/ios/EXUpdates/Multipart/EXUpdatesMultipartStreamReader.m:76]: shouldn't the stream close somewhere?

File: packages/expo-updates/ios/Tests/EXUpdatesMultipartStreamReaderTests.m [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+- (void)testSimpleCase
+{
+  NSString *response =
+  @"preamble, should be ignored\r\n"
+  @"--sample_boundary\r\n"
+  @"Content-Type: application/json; charset=utf-8\r\n"
+  @"Content-Length: 2\r\n\r\n"
+  @"{}\r\n"
+  @"--sample_boundary--\r\n"
+  @"epilogue, should be ignored";
```
Comment (ide) [inline packages/expo-updates/ios/Tests/EXUpdatesMultipartStreamReaderTests.m:25]: nice edge cases with preamble + epilogue

File: packages/expo-updates/ios/Tests/EXUpdatesFileDownloaderTests.m [hunk 13]
```diff
@@ -86,7 +76,29 @@ - (void)testExtraHeaders_OverrideOrder
...
+    }
+  }];
+  EXUpdatesFileDownloader *downloader = [[EXUpdatesFileDownloader alloc] initWithUpdatesConfig:config];
+
+  // assetRequestHeaders should not be able to override preset headers
+  NSDictionary *extraHeaders = @{
+    @"expo-platform": @"android"
+  };
+
+  NSURLRequest *actual = [downloader createGenericRequestWithURL:[NSURL URLWithString:@"https://u.expo.dev/00000000-0000-0000-0000-000000000000"] extraHeaders:extraHeaders];
```
Comment (ide) [inline packages/expo-updates/ios/Tests/EXUpdatesFileDownloaderTests.m:101]:
```suggestion
  NSURLRequest *actual = [downloader createGenericRequestWithURL:[NSURL URLWithString:@"https://assets.eascdn.net/00000000-0000-0000-0000-000000000000/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"] extraHeaders:extraHeaders];
```

**Review by ide — CHANGES_REQUESTED** [review]

**Comment by ide** [comment]
(Some of my review comments might be a bit outdated, sorry review took awhile.)

File: packages/expo-updates/ios/EXUpdates/AppLoader/EXUpdatesFileDownloader.m [hunk 14]
```diff
@@ -85,115 +88,280 @@ - (NSURLRequest *)createManifestRequestWithURL:(NSURL *)url extraHeaders:(nullab
...
+        } else if ([contentDispositionNameFieldValue isEqualToString:@"extensions"]) {
+          extensionsData = content;
+        }
+      }
+    }
+  } progressCallback:^(NSDictionary *headers, NSNumber *loaded, NSNumber *total) {}];
+  
+  if (!completed) {
+    NSError *error = [NSError errorWithDomain:EXUpdatesFileDownloaderErrorDomain
+                                         code:1044
```
[the remark this answers is not in the document]
Comment (wschurman) [reply packages/expo-updates/ios/EXUpdates/AppLoader/EXUpdatesFileDownloader.m:194]: https://linear.app/expo/issue/ENG-2676/constantenum-error-codes

File: packages/expo-updates/ios/EXUpdates/Multipart/EXUpdatesMultipartStreamReader.m (hunk 10 above)
Comment (wschurman) [reply packages/expo-updates/ios/EXUpdates/Multipart/EXUpdatesMultipartStreamReader.m:40]: yeah this one confused me for a bit as well. hesitant to change though since this class is a fork.

File: packages/expo-updates/ios/EXUpdates/Multipart/EXUpdatesMultipartStreamReader.m [hunk 15]
```diff
@@ -131,6 +130,7 @@ - (BOOL)readAllPartsWithCompletionCallback:(RCTMultipartCallback)callback
       delimiter = restDelimiter;
     }
   }
+  [_stream close];
```
[the remark this answers is not in the document]
Comment (wschurman) [reply packages/expo-updates/ios/EXUpdates/Multipart/EXUpdatesMultipartStreamReader.m:133]: Yeah, I think it's okay not to close it in general. It seems like most implementations just have it there as a formality/marker that prevents a subsequent call to `open`.

[the remark this answers is not in the document]
Comment (wschurman) [reply packages/expo-updates/ios/EXUpdates/Multipart/EXUpdatesMultipartStreamReader.m:133]: Will update this though.

**Review by ide — APPROVED** [review]
